src/shogun/classifier/Perceptron.cpp

```diff
@@ -78,44 +60,34 @@ bool CPerceptron::train_machine(CFeatures* data)
 		w.set_const(1.0 / num_feat);
 		bias=0;
 	}
-	else
-	{
-		w = get_w();
-	}
-	auto pb = SG_PROGRESS(range(max_iter));
-	//loop till we either get everything classified right or reach max_iter
-	while (!converged && iter < max_iter)
+}
+
+void CPerceptron::iteration()
+{
+	bool converged = true;
+	SGVector<float64_t> w = get_w();
+	SGVector<float64_t> output(features->get_num_vectors());
+
+	auto labels = binary_labels(m_labels)->get_int_labels();
```

Comment:
Suprisingly this works but doing .begin() in the same line does not.

Reply:
needs investigation!

Reply:
agreed